rx_signal_processing/decimate.cu

```diff
@@ -1,221 +1,693 @@
+ * filter that is zero padded to a power of two in length. This algorithm takes one or more 
+ * baseband datasets corresponding to each RX frequency and filters each one using a single lowpass
+ * filter before downsampling.
+ * 
+ *   gridDim.x - The number of decimated output samples for one frequency set.
+ *   gridDim.y - Total number of antennas.
+ *   gridDim.z - Total number of frequency data sets.
+ *
+ *   blockIdx.x - Decimated output sample index.
+ *   blockIdx.y - Antenna index.
+ *   blockIdx.z - Frequency dataset index.
+ *
+ *   blockDim.x - Number of filter taps in the lowpass filter.
+
+ *   threadIdx.x - Filter tap indices.
+ */
+__global__ void lowpass_decimate1024(cuComplex* original_samples,
+  cuComplex* decimated_samples,
+  cuComplex* filter_taps, uint32_t dm_rate,
+  uint32_t samples_per_antenna) {
+
+  // Since number of filter taps is calculated at runtime and we do not want to hardcode
+  // values, the shared memory can be dynamically initialized at invocation of the kernel.
+  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared
+
+  extern __shared__ cuComplex filter_products[];
+
+  auto frequency_dataset_offset = blockIdx.z * (gridDim.x * dm_rate);
+  
+  auto antenna_num = blockIdx.y;
+  auto antenna_offset = antenna_num * samples_per_antenna;
+
+  auto dec_sample_num = blockIdx.x;
+  auto dec_sample_offset = dec_sample_num * dm_rate;
+
+  auto tap_offset = threadIdx.x;
+
+  // If an offset should extend past the length of samples per antenna
+  // then zeroes are used as to not segfault or run into the next buffer.
+  // output samples using these will be discarded as to not introduce edge effects
+  cuComplex sample;
+  if ((dec_sample_offset + threadIdx.x) >= samples_per_antenna) {
+    sample = make_cuComplex(0.0f,0.0f);
+  }
+  else {
+    auto final_offset = frequency_dataset_offset + antenna_offset + dec_sample_offset + threadIdx.x;
+    sample = original_samples[final_offset];
+  }
+
+
+  filter_products[tap_offset] = cuCmulf(sample,filter_taps[tap_offset]);
+  // Synchronizes all threads in a block, meaning 1 output sample per rx freq
+  // is ready to be calculated with the parallel reduce
+  __syncthreads();
+
+  auto calculated_output_sample = parallel_reduce(filter_products, tap_offset);
+
+  // When decimating, we go from one set of samples for each antenna
+  // to multiple sets of reduced samples for each frequency. Output samples are
+  // grouped by frequency with all samples for each antenna following each other
+  // before samples of another frequency start.
+  if (threadIdx.x == 0) {
+    frequency_dataset_offset = blockIdx.z * gridDim.x;
+    antenna_offset = antenna_num * samples_per_antenna/dm_rate;
+    auto total_antennas = gridDim.y;
+    auto freq_offset = threadIdx.y * gridDim.x * total_antennas;
+    auto total_offset = frequency_dataset_offset + antenna_offset + freq_offset + dec_sample_num;
```

Comment:
We think that freq_offset and frequency_dataset_offset need to be combined here to be: blockIdx.z*gridDim.x*total_antennas